Add support for both report-only and content-security CSP headers

MartinPetkov opened the pull request:
Also refactor **init** to reduce complexity and fix `tox.ini` and tox test failures

Fixes https://github.com/sdelements/django-security/issues/44

funkaoshi commented on security/middleware.py:
```diff
@@ -655,42 +655,75 @@ def _csp_builder(self, csp_dict):
+                                  None)
+
         err_msg = 'Middleware requires either CSP_STRING or CSP_DICT setting'
+        report_err_msg = 'Middleware requires either CSP_REPORT_STRING, ' \
```
We prefer to do line continuation using brackets.

```
report_err_msg = (
    'Middleware requires either CSP_REPORT_STRING, '
    'CSP_REPORT_DICT setting, or neither. If neither, '
    ...    
)
```

funkaoshi commented on security/middleware.py:
```diff
@@ -655,42 +655,75 @@ def _csp_builder(self, csp_dict):
 
+    def _choose_csp_str(self, csp_dict, csp_str, err_msg):
+        if csp_dict and csp_str:
+            logger.warning('%s', err_msg)
```
Should log an error here, really.

funkaoshi commented on security/middleware.py:
```diff
@@ -655,42 +655,75 @@ def _csp_builder(self, csp_dict):
-            raise django.core.exceptions.MiddlewareNotUsed
+        if set_csp_str:
+            if not (csp_dict or csp_string):
+                logger.warning('%s, none found', err_msg)
```
Should log as an error here as well.

funkaoshi commented on security/middleware.py:
```diff
@@ -655,42 +655,75 @@ def _csp_builder(self, csp_dict):
+                self._csp_report_string = self._choose_csp_str(
+                    csp_report_dict,
+                    csp_report_string,
+                    report_err_msg)
```
Put the closing parenthesis on the next line is what we normally do.

funkaoshi commented on security/middleware.py:
```diff
@@ -655,42 +655,75 @@ def _csp_builder(self, csp_dict):
 
         return '; '.join(csp_components)
 
+    def _choose_csp_str(self, csp_dict, csp_str, err_msg):
```
Add docstring that says what csp_dict, csp_str are, what this is doing.

funkaoshi commented on security/middleware.py:
```diff
@@ -655,42 +655,88 @@ def _csp_builder(self, csp_dict):
     def __init__(self):
         # sanity checks
-        csp_mode = getattr(django.conf.settings, 'CSP_MODE', None)
+        self._csp_mode = getattr(django.conf.settings, 'CSP_MODE') or 'enforce'
```
You still need the 'None'. If you don't supply CSP_MODE getattr will raise an exception.

funkaoshi commented on security/middleware.py:
```diff
@@ -655,42 +655,88 @@ def _csp_builder(self, csp_dict):
+                                  None)
+
         err_msg = 'Middleware requires either CSP_STRING or CSP_DICT setting'
+        report_err_msg = (
```
Move this definition down to where it is used.

funkaoshi commented:
Why @MartinPetkov !

MartinPetkov commented:
It's just been a while and I thought it was forggoten and no one was watching it. I'm happy to reopen and contribute it.

MartinPetkov commented:
Rebased and updated. Tried running tox with Python 3.8, seems like only the Pep tests failed. PTAL.

Gee19 commented on testing/tests/tests.py:
```diff
@@ -988,6 +988,12 @@ def test_report_only_set(self):
 
+    def test_both_enforce_and_report_only(self):
+        with self.settings(CSP_MODE='enforce-and-report-only'):
+            response = self.client.get('/accounts/login/')
```
I think this line should grab `LOGIN_URL` from django settings.

We all miss you @MartinPetkov

MartinPetkov replied:
Done.

I also changed `tox.ini` so now all the tests pass, including pep8.

MartinPetkov replied:
Miss you too. Say hi to everyone!

Gee19 commented:
> Rebased and updated. Tried running tox with Python 3.8, seems like only the Pep tests failed. PTAL.

We need to stay pinned to python3.6 for the time being

MartinPetkov commented:
> > Rebased and updated. Tried running tox with Python 3.8, seems like only the Pep tests failed. PTAL.
> 
> We need to stay pinned to python3.6 for the time being

Done, reverted to Python 3.6. Tested with the `python:3.6` docker container.

MartinPetkov commented:
Ping?

sibartlett commented:
Sorry for the slow response, we’ve all been busy doing a two week hackathon. We’ll be starting regular sprint work again soon, so hopefully someone can take a look then 😄

MartinPetkov commented:
> Sorry for the slow response, we’ve all been busy doing a two week hackathon. We’ll be starting regular sprint work again soon, so hopefully someone can take a look then 😄

Sounds good, just making sure it didn't get lost forever. Hope the hackathon was fun!

Gee19 commented:
Gave this a quick test, thanks for cleaning up all the flake8 warnings as well.

<3 @MartinPetkov
